### ISSUE TITLE
Strict date parsing

### ISSUE THREAD

**Issue by micahellison** [jrnl-org/jrnl#630]
We have a variety of issues related to date parsing in v2, with the jrnl code interpreting data such as markdown (#443 and #614) and chat logs (#625) as dates simply because they are wrapped in square brackets. In rc3, we tried to tighten up the logic by ignoring anything less than 6 characters, but it's still not a comprehensive fix.

I think, in the perfect world, jrnl would parse dates strictly rather than loosely. That is, it would read the `timeformat` in the config file and use `strptime` with that format to parse it. If it can't be parsed, it's not treated as a date, and instead is just part of the journal entry.

However, this would make changing timeformats impossible without a conversion tool. If you wanted a new format and just changed it in your config file, jrnl would end up reading the whole thing as a single entry with today's date (at least given its current behavior). Not ideal.

There are also concerns related to Python versions - in [this SO thread](https://stackoverflow.com/questions/26165659/python-timezone-z-directive-for-datetime-strptime-not-available) it's said that strptime doesn't support the %z convention until 3.2, and that particular convention was raised in the bug #444.

So if we take this approach, we'd need to add:

- a command the user can enter to change the date format
- very clear documentation about NOT changing the date format in the config
- some live post-date-conversion tests to be absolutely sure the journal doesn't get nuked when changing the date format 
- some basic input validation on the new date format entered by the user (e.g. can it be converted to a Python date and back again with that format while ending up as the same date?)

There are some other, simpler approaches to this problem that I'll make separate issues for, but I think this could be a worthwhile long term goal for jrnl.

**Comment by zorab47** [comment]
It adds check complexity, but allowing multiple date formats to be specified in the config might be a solution. And each would be attempted before ignoring the value.

**Comment by peter-schmidbauer** [comment]
Using parsedatetime to parse the dates from the saved journal breaks a lot of custom date formats (i.e. the German format %d.%m.%Y) because it will parse them incorrectly (i.e. the day as month and vice versa). This is a clear regression and makes v2.0 unusable for me. 

Personally I don't think the journal config should be edited by hand (at all) and instead should be edited through commands (kind of like `git config`). That way, it would be possible to change the format later on.

Alternatively, at least use `strptime` by default everywhere and only fall back to parsedatetime when that doesn't work, as otherwise the custom date format is kinda pointless.

**Comment by wren** [comment]
@peter-schmidbauer I completely agree that users shouldn't have to edit the config manually. Our [v2.2 milestone](https://github.com/jrnl-org/jrnl/milestone/9)* focuses entirely on getting users out of the config. Every part of the config should be updatable through jrnl commands before by the end of that milestone.

* Please note that the milestone itself still needs to be fleshed out with more issues, but we do know what the end goal is.
